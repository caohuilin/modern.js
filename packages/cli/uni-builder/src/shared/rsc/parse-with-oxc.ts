import { logger } from '@modern-js/utils';
import { type FunctionBody, type ParseResult, parseAsync } from 'oxc-parser';

export const parseSourceWithOxc = async (source: string, filename: string) => {
  return await parseAsync(source, {
    sourceFilename: filename,
  });
};

interface ServerAction {
  localName: string;
  exportName?: string;
  span: {
    start: number;
    end: number;
  };
}

export const isClientModule = async (ast: ParseResult) => {
  return checkDirective(ast, 'use client');
};

const checkDirective = async (ast: ParseResult, originDirective: string) => {
  return ast.program.directives.find(
    directive => directive.directive === originDirective,
  );
};

// for 循环遍历 oxc 的 ast
// 判断当前模块有没有全局 use server 指令
// 如果有，收集所有 export 的函数，作为 Server Action
// 如果没有，收集声明 use server 的函数，作为 Server Action
// 检查每个函数有没有声明 use server 指令

// 当 hasModuleUseServer 为 false 时，所有最外层的函数都应该判断有没有 use server 指令
// 当 hasModuleUseServer 为 true 时，所有最外层的函数直接被收集

export const getModuleUseServerInfo = (ast: ParseResult) => {
  const directive = ast.program.directives.find(
    directive => directive.directive === 'use server',
  );
  if (!directive) return null;
  return {
    directive: directive.directive,
    start: directive.start,
    end: directive.end,
  };
};

export function getServerActions(ast: ParseResult): ServerAction[] {
  const hasModuleUseServer = getModuleUseServerInfo(ast);
  const serverActions: ServerAction[] = [];
  const currentExports: Map<string, string> = new Map();

  const checkUseServer = (body: FunctionBody) => {
    return body.directives.some(
      directive =>
        directive.type === 'Directive' && directive.directive === 'use server',
    );
  };
  for (const node of ast.program.body) {
    switch (node.type) {
      case 'ExportDefaultDeclaration':
        if (node.declaration) {
          if (node.declaration.type === 'FunctionDeclaration') {
            if (
              hasModuleUseServer ||
              (node.declaration.body && checkUseServer(node.declaration.body))
            ) {
              serverActions.push({
                localName: node.declaration.id?.name || '',
                exportName:
                  node.exported.type === 'Identifier'
                    ? node.exported.name
                    : 'default',
                span: {
                  start: node.start,
                  end: node.end,
                },
              });
            }
          }
          if (node.declaration.type === 'ArrowFunctionExpression') {
            if (
              hasModuleUseServer ||
              (node.declaration.body && checkUseServer(node.declaration.body))
            ) {
              serverActions.push({
                localName: 'default',
                exportName: 'default',
                span: {
                  start: node.start,
                  end: node.end,
                },
              });
            }
          }
          if (node.declaration.type === 'Identifier') {
            currentExports.set(node.declaration.name, 'default');
          }
        }
        break;
      case 'ExportNamedDeclaration':
        if (node.specifiers) {
          node.specifiers.forEach(specifier => {
            if (
              specifier.type === 'ExportSpecifier' &&
              specifier.local.type === 'Identifier' &&
              specifier.exported?.type === 'Identifier'
            ) {
              const localName = specifier.local.name;
              const exportName = specifier.exported.name;
              currentExports.set(localName, exportName);
            }
          });
        }
        if (node.declaration) {
          if (node.declaration.type === 'FunctionDeclaration') {
            if (
              hasModuleUseServer ||
              (node.declaration.body && checkUseServer(node.declaration.body))
            ) {
              serverActions.push({
                localName: node.declaration.id?.name || '',
                exportName: node.declaration.id?.name || '',
                span: {
                  start: node.start,
                  end: node.end,
                },
              });
            }
          } else if (node.declaration.type === 'VariableDeclaration') {
            for (const declarator of node.declaration.declarations) {
              if (
                declarator.init?.type === 'ArrowFunctionExpression' &&
                (hasModuleUseServer || checkUseServer(declarator.init.body))
              ) {
                serverActions.push({
                  localName:
                    declarator.id.type === 'Identifier'
                      ? declarator.id.name
                      : '',
                  exportName:
                    declarator.id.type === 'Identifier'
                      ? declarator.id.name
                      : '',
                  span: {
                    start: node.start,
                    end: node.end,
                  },
                });
              }
            }
          }
        }
        break;
      case 'FunctionDeclaration':
        if (hasModuleUseServer || (node.body && checkUseServer(node.body))) {
          serverActions.push({
            localName: node.id?.name || '',
            span: {
              start: node.start,
              end: node.end,
            },
          });
        }
        break;

      case 'VariableDeclaration':
        for (const declarator of node.declarations) {
          if (
            declarator.init?.type === 'ArrowFunctionExpression' &&
            (hasModuleUseServer || checkUseServer(declarator.init.body))
          ) {
            serverActions.push({
              localName:
                declarator.id.type === 'Identifier' ? declarator.id.name : '',
              span: {
                start: node.start,
                end: node.end,
              },
            });
          }
        }
        break;
    }
  }

  return serverActions.map(serverAction => ({
    ...serverAction,
    exportName:
      serverAction.exportName ?? currentExports.get(serverAction.localName),
  }));
}
