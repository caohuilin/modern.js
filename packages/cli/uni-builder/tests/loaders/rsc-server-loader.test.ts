import path from 'node:path';
import fs from 'fs/promises';
import { beforeEach, describe, expect, it, vi } from 'vitest';
import type { LoaderContext } from 'webpack';
import { MODERN_RSC_INFO, sharedData } from '../../src/shared/rsc/common';
import rscServerLoader, {
  type RscServerLoaderOptions,
} from '../../src/shared/rsc/rsc-server-loader';

async function callLoader(
  resourcePath: string,
  buildInfo?: Record<string, any>,
): Promise<string | Buffer> {
  const input = await fs.readFile(resourcePath);

  return new Promise((resolve, reject) => {
    const context: Partial<LoaderContext<RscServerLoaderOptions>> = {
      getOptions: () => ({}),
      resourcePath,
      cacheable: vi.fn(),
      _module: {
        buildInfo: buildInfo || {},
      } as any,
      async: () => context.callback!,
      callback: (error, content) => {
        if (error) {
          reject(error);
        } else if (typeof content === 'string') {
          resolve(content);
        } else {
          reject(
            new Error(`Did not receive any content from rscServerLoader.`),
          );
        }
      },
    };

    rscServerLoader.call(
      context as LoaderContext<RscServerLoaderOptions>,
      input.toString(`utf-8`),
      '',
    );
  });
}

describe('rscServerLoader', () => {
  beforeEach(() => {
    sharedData.clear();
  });

  it('support use client directive', async () => {
    const resourcePath = path.resolve(
      __dirname,
      `fixtures/client-component.jsx`,
    );
    const output = await callLoader(resourcePath);

    expect(output).toMatchSnapshot();
  });

  it(`should handle inline actions correctly`, async () => {
    const resourcePath = path.resolve(__dirname, `fixtures/inline-actions.tsx`);

    const output = await callLoader(resourcePath);

    expect(output).toMatchSnapshot();
  });

  it(`should handle server module correctly`, async () => {
    const resourcePath = path.resolve(__dirname, `fixtures/server-actions.ts`);

    const output = await callLoader(resourcePath);

    expect(output).toMatchSnapshot();
  });

  it('should generate client manifest correctly', async () => {
    const resourcePath = path.resolve(
      __dirname,
      `fixtures/client-component.jsx`,
    );

    const buildInfo: Record<string, any> = {};
    await callLoader(resourcePath, buildInfo);
    const { clientReferences } = buildInfo[MODERN_RSC_INFO];

    expect(clientReferences).toEqual([
      {
        id: 'tests/loaders/fixtures/client-component.jsx#ClassA',
        exportName: 'ClassA',
      },
      {
        id: 'tests/loaders/fixtures/client-component.jsx#ComponentA',
        exportName: 'ComponentA',
      },
      {
        id: 'tests/loaders/fixtures/client-component.jsx#MemoizedComponentA',
        exportName: 'MemoizedComponentA',
      },
      {
        id: 'tests/loaders/fixtures/client-component.jsx#ComponentB',
        exportName: 'ComponentB',
      },
      {
        id: 'tests/loaders/fixtures/client-component.jsx#foo',
        exportName: 'foo',
      },
      {
        id: 'tests/loaders/fixtures/client-component.jsx#ComponentC',
        exportName: 'ComponentC',
      },
      {
        id: 'tests/loaders/fixtures/client-component.jsx#ComponentD',
        exportName: 'ComponentD',
      },
      {
        id: 'tests/loaders/fixtures/client-component.jsx#bar',
        exportName: 'bar',
      },
      {
        id: 'tests/loaders/fixtures/client-component.jsx#ComponentE',
        exportName: 'ComponentE',
      },
      {
        id: 'tests/loaders/fixtures/client-component.jsx#ComponentF',
        exportName: 'ComponentF',
      },
      {
        id: 'tests/loaders/fixtures/client-component.jsx#default',
        exportName: 'default',
      },
    ]);
  });

  it('should generate serverReferencesMap correctly', async () => {
    const resourcePath = path.resolve(__dirname, `fixtures/server-actions.ts`);

    const buildInfo: Record<string, any> = {};
    await callLoader(resourcePath, buildInfo);
    const { exportNames } = buildInfo[MODERN_RSC_INFO];
    expect(exportNames).toEqual([`foo`, `bar`, `b`, `default`]);
  });
});
