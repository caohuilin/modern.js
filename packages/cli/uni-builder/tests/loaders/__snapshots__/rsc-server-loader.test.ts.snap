// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`rscServerLoader > should handle inline actions correctly 1`] = `
"
import { registerServerReference } from \\"react-server-dom-webpack/server\\";/ @ts-ignore
import { z } from 'zod';

export async function foo() {
  'use server';

  return \`foo\`;
}
registerServerReference(foo, module.id, \\"foo\\");

export async function bar() {
  return qux();
}

const b = () => {
  'use server';

  return \`baz\`;
};
registerServerReference(b, module.id, \\"baz\\");

export { b as baz };

async function qux() {
  'use server';

  return \`qux\`;
}
registerServerReference(qux, module.id, \\"qux\\");

export default function () {
  'use server';

  return \`default\`;
}
registerServerReference(, module.id, \\"default\\");
"
`;

exports[`rscServerLoader > should handle server module correctly 1`] = `
"'use server';
import { registerServerReference } from \\"react-server-dom-webpack/server\\";
export async function foo() {
  return 'foo';
}
registerServerReference(foo, module.id, \\"foo\\");

export const bar = async () => 'bar';
registerServerReference(bar, module.id, \\"bar\\");

const baz = async () => 'baz';
registerServerReference(baz, module.id, \\"b\\");

export { baz as b };

export const qux = 'qux';

async function qoo() {
  return 'qoo';
}
registerServerReference(qoo, module.id, \\"default\\");

export default qoo;
"
`;

exports[`rscServerLoader > support use client directive 1`] = `
"
'use client';

import { registerClientReference } from \\"react-server-dom-webpack/server\\";
function createClientReferenceProxy(exportName) {
  return () => {
    throw new Error(\`Attempted to call \${exportName}() from the server but \${exportName} is on the client. It's not possible to invoke a client function from the server, it can only be rendered as a Component or passed to props of a Client Component.\`);
  };
}
export const ClassA = registerClientReference(createClientReferenceProxy(\\"ClassA\\"), \\"tests/loaders/fixtures/client-component.jsx#ClassA\\", \\"ClassA\\");
export const ComponentA = registerClientReference(createClientReferenceProxy(\\"ComponentA\\"), \\"tests/loaders/fixtures/client-component.jsx#ComponentA\\", \\"ComponentA\\");
export const MemoizedComponentA = registerClientReference(createClientReferenceProxy(\\"MemoizedComponentA\\"), \\"tests/loaders/fixtures/client-component.jsx#MemoizedComponentA\\", \\"MemoizedComponentA\\");
export const ComponentB = registerClientReference(createClientReferenceProxy(\\"ComponentB\\"), \\"tests/loaders/fixtures/client-component.jsx#ComponentB\\", \\"ComponentB\\");
export const foo = registerClientReference(createClientReferenceProxy(\\"foo\\"), \\"tests/loaders/fixtures/client-component.jsx#foo\\", \\"foo\\");
export const ComponentC = registerClientReference(createClientReferenceProxy(\\"ComponentC\\"), \\"tests/loaders/fixtures/client-component.jsx#ComponentC\\", \\"ComponentC\\");
export const ComponentD = registerClientReference(createClientReferenceProxy(\\"ComponentD\\"), \\"tests/loaders/fixtures/client-component.jsx#ComponentD\\", \\"ComponentD\\");
export const bar = registerClientReference(createClientReferenceProxy(\\"bar\\"), \\"tests/loaders/fixtures/client-component.jsx#bar\\", \\"bar\\");
export const ComponentE = registerClientReference(createClientReferenceProxy(\\"ComponentE\\"), \\"tests/loaders/fixtures/client-component.jsx#ComponentE\\", \\"ComponentE\\");
export const ComponentF = registerClientReference(createClientReferenceProxy(\\"ComponentF\\"), \\"tests/loaders/fixtures/client-component.jsx#ComponentF\\", \\"ComponentF\\");
export default registerClientReference(() => {
  throw new Error(\\"Attempted to call the default export of <ROOT>/tests/loaders/fixtures/client-component.jsx from the server but it's on the client. It's not possible to invoke a client function from the server, it can only be rendered as a Component or passed to props of a Client Component.\\");
}, \\"tests/loaders/fixtures/client-component.jsx#default\\", \\"\\");"
`;
